pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -550,9 +551,89 @@ func makeShardedDescriptor(b BuildCtx, t alterPrimaryKeySpec) *catpb.ShardedDesc
 func recreateAllSecondaryIndexes(
```

Comment:
This new method reads all good to me. But have you consulted how the legacy schema changer handled secondary index rewrite: https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/sql/alter_primary_key.go#L645-L659 to make sure there is no behavioral discrepancy? There, the idea is to lay out conditions for secondary indexes that needs rewriting, but here it seems to me we rewrite *all* secondary indexes.

Reply:
Can you go into more detail, please? I'm not sure what you mean. This code you linked to refers to maintaining the UNIQUE property on the columns in the old primary key by creating an extra secondary index. This is a distinct consideration from the need to rewrite existing secondary indexes (which is largely motivated by the change in key suffix columns).

Reply:
I think @Xiang-Gu meant this : http://github.com/cockroachdb/cockroach/blob/f08a1b00f4d28946288fb0a20fe4be7b6861f10c/pkg/sql/alter_primary_key.go#L389-L389
though I think we only avoid rewrite in some very corner cases of unique index, I'd prefer not to copy this complex logic over :)

Reply:
I think that I just assumed that this logic was covered by existing tests. I didn't look too closely into what the legacy schema changer did, that's for sure. I'll look into this more, at least to make sure that these tests which I assumed existed actually do exist, and if not, add them (along with the accompanying logic in the declarative schema changer).

Reply:
I've thought about this and I'm not sure that it's worth dealing with these edge cases. The impact is going to be marginal.

Reply:
agreed, avoiding rewrites only for that very corner corner cases only introduce complexity without much gain in general.